lib/stores/S3Store.ts

```diff
@@ -374,6 +356,48 @@ export default class S3Store extends DataStore {
     this.cache.delete(id)
   }
 
+  private calcOptimalPartSize(size: number): number {
+    let optimalPartSize: number
+
+    // When upload is smaller or equal to PreferredPartSize, we upload in just one part.
+    if (size <= this.preferredPartSize) {
+      optimalPartSize = size
+    }
+    // Does the upload fit in MaxMultipartParts parts or less with PreferredPartSize.
+    else if (size <= this.preferredPartSize * this.maxMultipartParts) {
+      optimalPartSize = this.preferredPartSize
+      // Prerequisite: Be aware, that the result of an integer division (x/y) is
+      // ALWAYS rounded DOWN, as there are no digits behind the comma.
+      // In order to find out, whether we have an exact result or a rounded down
+      // one, we can check, whether the remainder of that division is 0 (x%y == 0).
+      //
+      // So if the result of (size/MaxMultipartParts) is not a rounded down value,
+      // then we can use it as our optimalPartSize. But if this division produces a
+      // remainder, we have to round up the result by adding +1. Otherwise our
+      // upload would not fit into MaxMultipartParts number of parts with that
+      // size. We would need an additional part in order to upload everything.
+      // While in almost all cases, we could skip the check for the remainder and
+      // just add +1 to every result, but there is one case, where doing that would
+      // doom our upload. When (MaxObjectSize == MaxPartSize * MaxMultipartParts),
+      // by adding +1, we would end up with an optimalPartSize > MaxPartSize.
+      // With the current S3 API specifications, we will not run into this problem,
+      // but these specs are subject to change, and there are other stores as well,
+      // which are implementing the S3 API (e.g. RIAK, Ceph RadosGW), but might
+      // have different settings.
+    } else if (size % this.maxMultipartParts === 0) {
+      optimalPartSize = size / this.maxMultipartParts
```

Comment:
I think this division is ok, because we check before is size is multiple of maxMultipartParts.